src/axom/primal/tests/primal_zip.cpp

```diff
@@ -0,0 +1,282 @@
+// Copyright (c) 2017-2021, Lawrence Livermore National Security, LLC and
+// other Axom Project Developers. See the top-level LICENSE file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+#include "axom/config.hpp"
+
+#include "axom/core/execution/for_all.hpp"
+#include "axom/core/memory_management.hpp"
+
+#include "axom/primal/geometry/Point.hpp"
+#include "axom/primal/utils/ZipIndexable.hpp"
+#include "axom/primal/utils/ZipPoint.hpp"
+#include "axom/primal/utils/ZipVector.hpp"
+#include "axom/primal/utils/ZipBoundingBox.hpp"
+#include "axom/primal/utils/ZipRay.hpp"
+
+#include "gtest/gtest.h"
+
+using namespace axom;
+
+//------------------------------------------------------------------------------
+
+template <typename ExecSpace, typename PrimitiveType>
+void check_zip_points_3d()
+{
+  using ZipType = primal::ZipIndexable<PrimitiveType>;
+
+  const int current_allocator = axom::getDefaultAllocatorID();
+  axom::setDefaultAllocator(axom::execution_space<ExecSpace>::allocatorID());
+
+  // create arrays of data
+  constexpr int N = 8;
+  double* x = axom::allocate<double>(N);
+  double* y = axom::allocate<double>(N);
+  double* z = axom::allocate<double>(N);
+
+  bool* valid = axom::allocate<bool>(N);
+
+  axom::for_all<ExecSpace>(
+    0,
+    8,
+    AXOM_LAMBDA(int idx) {
+      x[idx] = (idx % 2) + 1;
+      y[idx] = ((idx / 2) % 2) + 1;
+      z[idx] = (idx / 4) + 1;
+    });
+
+  ZipType it {{x, y, z}};
+
+  axom::for_all<ExecSpace>(
+    0,
+    8,
+    AXOM_LAMBDA(int idx) {
+      valid[idx] = (it[idx] == PrimitiveType {x[idx], y[idx], z[idx]});
+    });
+
+  for(int i = 0; i < N; i++)
+  {
+    EXPECT_EQ(valid[i], true);
+  }
+
+  axom::deallocate(x);
+  axom::deallocate(y);
+  axom::deallocate(z);
+  axom::deallocate(valid);
+  axom::setDefaultAllocator(current_allocator);
+}
+
+template <typename ExecSpace>
+void check_zip_bbs_3d()
+{
+  using BoxType = primal::BoundingBox<double, 3>;
+  using PointType = typename BoxType::PointType;
+  using ZipType = primal::ZipIndexable<BoxType>;
```

Comment:
This is really nice!

I am not 100% sure if the following is also necessary, but I strongly suspect it might be.

Can we also support an example of 2D points using 3 input arrays, the last of which is a `nullptr`.

E.g. something like:
```cpp
  // Setup: 3D input

  using ZipType = primal::ZipIndexable<Box3D>; // ??
  constexpr int N = 8;

  double* xmin = axom::allocate<double>(N);
  double* ymin = axom::allocate<double>(N);
  double* zmin = nullptr;

  double* xmax = axom::allocate<double>(N);
  double* ymax = axom::allocate<double>(N);
  double* zmax = nullptr;

  ZipType it {{xmin, ymin, zmin}, {xmax, ymax, zmax}};

  ....

  // Access as 2D entities: 
  // I'm not entirely sure what this should look like
  Box2D box2d_a = it.get<Box2D>(5);  // maybe?
  Box2D box2d_b = it.get<2>(5);          // maybe?
```

The use case is that the host code is not templated on the dimension and supports multiple dimensions, but the internal axom parts are templated.

Perhaps this can be added in a follow-up PR?